File: Difficulty: Medium/Case-specific Sorting of Strings/casespecific-sorting-of-strings.cpp
class Solution {
  public:
    string caseSort(string& s) {
        int n = s.size();

        int lower[26] = { 0 };
        int upper[26] = { 0 };

        // 1) Count frequencies
        for (char ch : s) {
            if (islower(ch)) {
                lower[ch - 'a']++;
            } else {
                upper[ch - 'A']++;
            }
        }

        // 2) Reconstruct result by consuming counts
        string result = s;
        int    l      = 0; // index for lowercase freq array
        int    u      = 0; // index for uppercase freq array

        for (int i = 0; i < n; i++) {
            if (islower(s[i])) {
                while (lower[l] == 0) {
                    l++;
                }
                result[i] = 'a' + l;
                lower[l]--;
            } else {
                while (upper[u] == 0) {
                    u++;
                }
                result[i] = 'A' + u;
                upper[u]--;
            }
        }

        return result;
    }
};

File: Difficulty: Medium/Case-specific Sorting of Strings/casespecific-sorting-of-strings.java
class Solution {
    public static String caseSort(String s) {
        int n = s.length();

        List<Character> lower = new ArrayList<>();  // to hold all lowercase letters
        List<Character> upper = new ArrayList<>();  // to hold all uppercase letters

        // 1) Partition into lowercase and uppercase lists
        for (char ch : s.toCharArray()) {
            if (Character.isLowerCase(ch)) {
                lower.add(ch);
            } else {
                upper.add(ch);
            }
        }

        // 2) Sort both lists independently
        Collections.sort(lower);
        Collections.sort(upper);

        // 3) Reconstruct result by consuming from sorted lists
        StringBuilder result = new StringBuilder(n);
        int           l      = 0; // index for lowercase list
        int           u      = 0; // index for uppercase list

        for (char ch : s.toCharArray()) {
            if (Character.isLowerCase(ch)) {
                result.append(lower.get(l++));  // place next sorted lowercase
            } else {
                result.append(upper.get(u++));  // place next sorted uppercase
            }
        }

        return result.toString();
    }
}